### ISSUE TITLE
Mate in one : M4 ?

### ISSUE THREAD

**Issue by tissatussa** [sictransit/woodpusher#31]
Until now i didn't manage to win a 5m+3s game against v1.2.1, but some engines around rating 1400 can .. at this moment it's not clear to me how strong your newest version is.

In one of the engine-vs-engine games this happened :

8/5kp1/p1N1p2p/8/3Pn1K1/6Pr/8/5q2 b - - 1 45

![M1-but-M4](https://github.com/user-attachments/assets/75ccad96-fe5a-412a-8e4e-c3daf0de5b18)

Woodpusher shows -M4 here but it's -M1 : Qf5#
What's wrong ?

Btw. the opponent engine, Minimardy v1.3 (CCRL blitz rating ~1342), uses the XBoard protocol so it can't display a mate like "M#" but .. it did NOT find the mate either !? Is this a confusing position for both engines ? It's weird ..

Another thing : this game also shows me that Minimardy should negate its eval value. For this, CuteChess (GUI) has the option "Scores from white's perspective", but i didn't set that checkbox yet.

```
[Event "engine-vs-engine"]
[Site "Holland"]
[Date "2024.11.28"]
[Round "?"]
[White "Woodpusher v1.2.1"]
[Black "Minimardi v1.3"]
[Result "0-1"]
[ECO "A04"]
[GameDuration "00:11:02"]
[Opening "Reti Opening"]
[PlyCount "94"]
[TimeControl "300+3"]

1. Nf3 {+0.28/7 4.6s} c5 2. e4 {+0.24/6 1.8s} Nc6 3. Nc3 {+0.34/6 2.6s} e6
4. Bd3 {+0.31/6 2.8s} Nf6 {+0.26/9 15s} 5. O-O {+0.21/7 8.6s} Bd6 {+0.33/8 15s}
6. Re1 {+0.34/6 3.4s} e5 {+0.23/9 13s} 7. Nd5 {+0.28/6 6.2s} O-O {+0.23/9 11s}
8. c3 {+0.22/6 3.4s} a6 {+0.22/9 9.8s} 9. a3 {+0.35/5 2.2s} b5 {+0.22/9 8.9s}
10. Bc2 {+0.17/6 5.6s} c4 {+0.30/8 8.2s} 11. d4 {+0.58/6 4.4s}
exd4 {+0.36/9 8.0s} 12. cxd4 {+0.65/6 2.7s} Nxd5 {+0.25/10 8.0s}
13. exd5 {+0.49/7 7.1s} Ne7 {+0.21/9 7.8s} 14. Ng5 {+0.22/6 3.6s}
f5 {+0.20/8 7.8s} 15. a4 {+0.35/6 4.4s} Qa5 {-0.28/7 7.7s}
16. Be3 {+0.09/6 6.5s} b4 {-0.45/8 7.6s} 17. Rc1 {-0.19/6 8.7s}
b3 {-0.59/7 7.5s} 18. Bb1 {-0.58/7 10s} Bb7 {-0.97/8 7.4s}
19. Qh5 {-0.62/6 4.5s} h6 {-1.04/8 7.3s} 20. Nf3 {-0.79/7 13s}
Bxd5 {-1.20/8 7.2s} 21. Rcd1 {-0.97/7 13s} Qxa4 {-1.48/8 7.1s}
22. Ne5 {-1.02/6 3.5s} Bb4 {-1.56/7 7.0s} 23. Bd2 {-1.10/7 14s}
Bxd2 {-1.46/8 7.0s} 24. Rxd2 {-1.23/7 10s} Be6 {-1.48/7 6.9s}
25. Rde2 {-0.99/6 4.6s} Nd5 {-1.34/8 6.8s} 26. Ng6 {-0.81/7 6.9s}
Rf6 {-1.60/8 6.7s} 27. Re5 {-0.94/7 6.8s} Qb4 {-1.56/8 6.6s}
28. Rf1 {-0.91/6 10s} c3 {-1.66/8 6.6s} 29. Bxf5 {-1.46/8 19s}
Rc8 {-1.27/8 6.5s} 30. Bxe6+ {-0.82/8 19s} dxe6 {-2.27/9 6.4s}
31. bxc3 {-1.49/8 20s} Rxc3 {-2.27/8 6.5s} 32. Qg4 {-1.83/7 14s}
b2 {-2.78/8 6.6s} 33. Ree1 {-2.48/7 7.5s} Rb3 {-3.70/8 6.7s}
34. Rb1 {-3.60/7 7.3s} Nc3 {-3.90/8 6.8s} 35. Rbe1 {-4.61/7 7.6s}
b1=Q {-10.30/9 6.8s} 36. Rxb1 {-7.39/8 8.6s} Rxb1 {-10.30/9 7.0s}
37. Ne7+ {-7.82/9 11s} Kf7 {-10.39/10 7.1s} 38. g3 {-7.84/9 15s}
Rxf1+ {-10.46/10 7.2s} 39. Kxf1 {-8.04/9 14s} Qb1+ {-11.37/10 7.3s}
40. Kg2 {-7.68/7 1.5s} Ne4 {-12.93/10 7.5s} 41. Nc6 {-7.18/7 2.5s}
Rxf2+ {-7.91/14 7.6s} 42. Kh3 {-M8/8 1.5s} Qf1+ {-7.91/12 7.8s}
43. Kh4 {-M6/9 2.7s} Rxh2+ {-7.91/13 8.0s} 44. Qh3 {-M4/8 0.89s}
Rxh3+ {-7.91/2 1.6s} 45. Kg4 {-M4/8 1.0s} Rxg3+ {-7.91/13 8.8s}
46. Kh5 {-M4/8 1.8s} Rh3+ {1.2s} 47. Kg4 {-M4/8 1.2s}
Qf5# {-7.91/8 0.022s, Black mates} 0-1
```

**Comment by sictransit** [comment]
This has now been fixed (among other things) in my current branch. It'll go into main soon as v1.3.0.

```
position fen 8/5kp1/p1N1p2p/8/3Pn1K1/6Pr/8/5q2 b - - 1 45
go
info depth 1 seldepth 1 nodes 40 nps 2500 hashfull 0 score cp 1791 time 22 pv h3g3
info depth 2 seldepth 2 nodes 163 nps 5620 hashfull 0 score mate 1 time 32 pv h6h5
info string debug aborting search @ depth 2, mate in 1
bestmove h6h5
```

**Comment by tissatussa** [comment]
> This has now been fixed (among other things) in my current branch. It'll go into main soon as v1.3.0

Nice ! I will try this version.

Regarding v1.2.1 : i let it play several 5m+3s games and now and then it made an illegal move, here are 5 examples. And here's the ZIPped log of the first one : [woodpusher-illegal-move-log.zip](https://github.com/user-attachments/files/17965346/woodpusher-illegal-move-log.zip)

![woodpusher-v1 2 1-illegal-move-A](https://github.com/user-attachments/assets/a86d4061-dc97-450f-a372-2ca44e8e0725)

![woodpusher-v1 2 1-illegal-move-B](https://github.com/user-attachments/assets/6e7d89c9-2f07-4ad5-8fc8-06888e94790b)

![woodpusher-v1 2 1-illegal-move-C](https://github.com/user-attachments/assets/62d30c05-065d-4b5e-bc7c-488f035bab20)

![woodpusher-v1 2 1-illegal-move-D](https://github.com/user-attachments/assets/45e3278e-70e3-4915-a977-9b2a113c5f10)

![woodpusher-v1 2 1-illegal-move-E](https://github.com/user-attachments/assets/c09d9c1a-b500-4fe6-8f2a-a4dc80dada5c)

Maybe this illegal move issue is also solved with v1.3.0 ?
